### PR TITLE
bpf_host: emit '-> network' traces for egress packets

### LINKED ISSUE #12561: Monitor Aggregation for connections does not work with host firewall
The host firewall does not respect the `monitor` parameter to the `ct_lookup[46]()` functions, it simply gathers it and discards, for example:

https://github.com/cilium/cilium/blob/f55ec9066e3a23461a0d14e20b3be8ab49a04c97/bpf/lib/host_firewall.h#L268
https://github.com/cilium/cilium/blob/f55ec9066e3a23461a0d14e20b3be8ab49a04c97/bpf/lib/host_firewall.h#L300-L301

This means that even with `monitorAggregationLevel=medium`, aggregation of monitor notifications based on connections does not occur correctly for packets from the host towards the networks, meaning that notifications are generated once per packet rather than once every 5s per connection.

The `monitor` argument should be handled similarly to how it is used in `bpf/bpf_lxc.c` in the `bpf_host.c`.

### PR DESCRIPTION
**Pull request by navarrothiago**
bpf_host: emit '-> network' traces for egress packets

See the issue below for the full context. The caveat is, this program is only attached to the native device egress if the host firewall is enabled, so there still won't be any TRACE_TO_NETWORK visibility if that's not the case.

I've also added handle_to_netdev_ipv4() to clean up to_netdev() a bit, to bring it on par with the ipv6 handling.

bpf_host: emit '-> network' traces for packets leaving the node when Host Firewall are enabled

Based on: #13485
Fixes: #12561

Co-authored-by: Thiago Navarro <navarro@accuknox.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Thiago Navarro <navarro@accuknox.com>

### CODE REVIEW

File: bpf/lib/common.h [hunk 1]
```diff
@@ -563,31 +563,37 @@ static __always_inline __u32 or_encrypt_key(__u8 key)
...
 	CB_CT_STATE,
-#define	CB_ADDR_V6_4		CB_CT_STATE	/* Alias, non-overlapping */
-#define	CB_ENCRYPT_DST		CB_CT_STATE	/* Alias, non-overlapping,
-						 * Not used by xfrm.
-						 */
-#define	CB_CUSTOM_CALLS		CB_CT_STATE	/* Alias, non-overlapping */
+#define	CB_ADDR_V6_4		CB_CT_STATE
+#define	CB_ENCRYPT_DST		CB_CT_STATE
+#define	CB_CUSTOM_CALLS		CB_CT_STATE
+#define	CB_CT_MONITOR		CB_CT_STATE
```
Comment (pchaigno) [inline bpf/lib/common.h]: What happened in this file?

File: bpf/lib/trace.h [hunk 2]
```diff
@@ -141,6 +141,13 @@ static __always_inline bool emit_trace_notify(__u8 obs_point, __u32 monitor)
 		}
 	}
 
+	/*
+	 * Ignore sample when aggregation is enabled and 'monitor' is set to 0.
+	 * Rate limiting (trace message aggregation) relies on connection tracking,
+	 * so if there is no CT information available at the observation point,
+	 * then 'monitor' will be set to 0 to avoid emitting trace notifications
+	 * when aggregation is enabled (the default).
+	 */
```
Comment (pchaigno) [inline bpf/lib/trace.h:150]: This sort of cleanup should be in a separate commit.

**Review by pchaigno — COMMENTED** [review]
This is a bit hard to review as is. Could you separate in multiple logical commits? For example, the changes to take monitor into account in existing traces (what #12561 is about) could be one commit, new traces would be another commit, and general cleanups (e.g., new comments) would be a last commit.

**Comment by navarrothiago** [comment]
> new traces would be another commit

Thank you for you comments. Sorry, but I did not understand. What do you mean by new traces? AFAIU, the issue is related to this [comment](https://github.com/cilium/cilium/issues/12561#issuecomment-705089371), right? So, `new trace` would be the `issue`. Let me know if I got your point.

File: bpf/lib/nodeport.h [hunk 3]
```diff
@@ -1217,13 +1222,17 @@ static __always_inline int nodeport_nat_ipv4_fwd(struct __ctx_buff *ctx)
...
+	__u32 monitor = ctx_load_meta(ctx, CB_CT_MONITOR);
 
 	if (snat_v4_needed(ctx, &target.addr, &from_endpoint))
 		ret = snat_v4_process(ctx, NAT_DIR_EGRESS, &target,
 				      from_endpoint);
+
 	if (ret == NAT_PUNT_TO_STACK)
 		ret = CTX_ACT_OK;
 
+	send_trace_notify(ctx, TRACE_TO_NETWORK, 0, 0, 0, 0, ret, monitor);
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: I think we should skip the commit that adds `TRACE_TO_NETWORK` traces for now. See the discussion at https://github.com/cilium/cilium/pull/14696. These additional traces will get very noisy quickly so there's a bit more work needed before we can introduce them.

**Review by pchaigno — CHANGES_REQUESTED** [review]
I think we should exclude the second commit for now. The other two commits look :+1: to me.

Please don't forget to add Timo has a coauthor of the last commit (via `Signed-off-by`, `Co-authored-by`, or both).

**Comment by pchaigno** [comment]
@navarrothiago Could you also fix the second commit's title? Right now, I doesn't provide any useful/actionable information.

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Comment by navarrothiago** [comment]
@pchaigno, should we re-trigger the jobs?

**Comment by pchaigno** [comment]
test-runtime

**Comment by pchaigno** [comment]
Team reviews are covered and tests are passing. Marking as ready to merge.

**Pull request merged**